includes/utils/class-amp-validation-utils.php

```diff
@@ -231,14 +301,133 @@ public static function validate_content( $post ) {
+		foreach ( $pending_wrap_callbacks as $hook => $callbacks ) {
+			foreach ( $callbacks as $callback ) {
+				remove_action( $hook, $callback['function'], $callback['priority'] );
+				$wrapped_callback = self::wrapped_callback( $callback );
+				add_action( $hook, $wrapped_callback, $callback['priority'], $callback['accepted_args'] );
+			}
+		}
+	}
+
+	/**
+	 * Gets the plugin of the callback, if one exists.
+	 *
+	 * @param string|array $callback The callback for which to get the plugin.
+	 * @return string|null $plugin   The plugin to which the callback belongs, or null.
+	 */
+	public static function get_plugin( $callback ) {
+		try {
+			if ( is_string( $callback ) && is_callable( $callback ) ) {
+				// The $callback is a function or static method.
+				$exploded_callback = explode( '::', $callback );
+				if ( count( $exploded_callback ) > 1 ) {
+					$reflection = new ReflectionClass( $exploded_callback[0] );
+				} else {
+					$reflection = new ReflectionFunction( $callback );
+				}
+			} elseif ( is_array( $callback ) && isset( $callback[0], $callback[1] ) && method_exists( $callback[0], $callback[1] ) ) {
+				// The $callback is a method.
+				$reflection = new ReflectionClass( $callback[0] );
+			} elseif ( is_object( $callback ) && ( 'Closure' === get_class( $callback ) ) ) {
+				$reflection = new ReflectionFunction( $callback );
+			}
+
+			$file = isset( $reflection ) ? $reflection->getFileName() : null;
+			if ( ! isset( $file ) ) {
+				return null;
+			}
+			$source_data = self::get_source( $file );
+			if ( 'plugins' === $source_data['type'] ) {
+				return $source_data['source'];
+			}
+		} catch ( Exception $e ) {
+			return null;
+		}
+		return null;
+	}
+
+	/**
+	 * Wraps a callback in comments if it outputs markup.
+	 *
+	 * If the sanitizer removes markup,
+	 * this indicates which plugin it was from.
+	 * The call_user_func_array() logic is mainly copied from WP_Hook:apply_filters().
+	 *
+	 * @param array $callback {
+	 *     The callback data.
+	 *
+	 *     @type callable $function
+	 *     @type int      $accepted_args
+	 *     @type string   $plugin
+	 * }
+	 * @return closure $wrapped_callback The callback, wrapped in comments.
+	 */
+	public static function wrapped_callback( $callback ) {
+		return function() use ( $callback ) {
+			$function      = $callback['function'];
+			$accepted_args = $callback['accepted_args'];
+			$args          = func_get_args();
+
+			ob_start();
+			if ( 0 === $accepted_args ) {
+				$result = call_user_func_array( $function, array() );
+			} elseif ( $accepted_args >= func_num_args() ) {
+				$result = call_user_func_array( $function, $args );
+			} else {
+				$result = call_user_func_array( $function, array_slice( $args, 0, intval( $accepted_args ) ) );
+			}
+			$output = ob_get_clean();
+
+			if ( ! empty( $output ) ) {
+				printf( '<!--before:%s-->', esc_attr( $callback['plugin'] ) );
```

Comment:
I think the source type should also be included in these comments. So it should be like `<!--before:plugin:jetpack-->`, for example. This would help in the case where a theme and plugin have the same name.

Reply:
Sure, [this commit](https://github.com/Automattic/amp-wp/commit/76ec4dfc770305a0624e16ffe9ec0ee84729d732) adds the type of source to the comments, like you suggested.